### DISCUSSION TITLE
Displaying offline map - load mbtiles file

### DISCUSSION THREAD

**Discussion by urossavelj** [Mapsui/Mapsui#2182] [Q&A]
Hi,

so after the previous issue with getting the extent of the visible map now I have another one. I loaded the visible tiles to a local sqlite db. I would now like to display the map from the tiles loaded for ex. when a button is clicked.

I tried like this but I get just an empty view
```
var tileSource = new MbTilesTileSource(new SQLiteConnectionString(Path.Combine(FileSystem.AppDataDirectory, "myOfflineDb.mbtiles")));

        _mapControl = new Mapsui.UI.Maui.MapControl();
        _mapControl.Map?.Layers.Add(new TileLayer(tileSource));

        Content = _mapControl;
```
I checked the myOfflineDb.mbtiles file with DB Browser for SQLite and it has the tiles inside, so that's ok. Is there anything else I have to set for it to display the tiles? What am I missing here?

**Comment by charlenni** [comment]
Do you moved around? Is there no tile at all?

**Reply by urossavelj** [reply]
Yes, I tried to move around but it's totally empty, just blank space

**Reply by urossavelj** [reply]
I tried putting the base world.mbtiles there and read it with the same method and that works. So something must be wrong with the tiles that I save in the sqlite database, but I have no idea what

**Reply by charlenni** [reply]
The mbtiles file is a normal SQLite database. Try to open both with a tool and compare them. They need the same format.

**Reply by charlenni** [reply]
Did you enabled logging for Mapsui? Sometimes some errors are only displayed when you log them.

**Reply by urossavelj** [reply]
This is the database of the tiles I loaded manually:
![image](https://github.com/Mapsui/Mapsui/assets/14927094/7ed3a012-2223-4441-8ef1-f35d5dd144e3)

tiles table:
![image](https://github.com/Mapsui/Mapsui/assets/14927094/2e849de9-ec68-4995-881b-1bf502b6dcfa)

And this is the database, that is default called "world.mbtiles":
![image](https://github.com/Mapsui/Mapsui/assets/14927094/1fd5c16c-743e-408a-98ee-6a862d667629)

It looks to have much more tables inside.

I save the tiles in the database with the logic from BruTiles example:
https://github.com/BruTile/BruTile

This is actually what I'm doing:
```
        var tileInfos = tileSource.Schema.GetTileInfos(extent, resolution);

        Debug.WriteLine("Show tile info");
        foreach (var tileInfo in tileInfos)
        {
            var tile = await tileSource.GetTileAsync(tileInfo);

            Debug.WriteLine(
                $"tile col: {tileInfo.Index.Col}, " +
                $"tile row: {tileInfo.Index.Row}, " +
                $"tile level: {tileInfo.Index.Level} , " +
                $"tile size {tile.Length}");

            mbStreetTileCache.Add(tileInfo.Index, tile);
        }
```

Add method is from this example https://github.com/spaddlewit/MBTilesPersistentCache/blob/main/MBTilesPersistentCache/MbTileCache.cs#L48

**Reply by inforithmics** [reply]
Have you only stored tiles on zoom level 8? If yes then it only works below this zoom level. Because mapsui looks only for higher zoom levels ( Level < 8) but not for lower zoom levels ( level > 8).

**Reply by urossavelj** [reply]
Yes, I zoomed on the map and than stored tiles of the zoomed area. Now I would like to display that part of the map

**Reply by charlenni** [reply]
What @inforithmics asked is, did you while you display the offline tiles, you zoom in to zoom level 8 or higher? If you have zoom level 7 or lower, you will see nothing.

**Reply by urossavelj** [reply]
Ah, I understand what you meant now. I also tried zooming the map after I load the offline tiles, but it's still the same.

**Reply by inforithmics** [reply]
Is it necessary to use the MbTileCache? An alternative would be
https://github.com/Mapsui/Mapsui/blob/master/Mapsui.Extensions/Cache/SqlitePersistentCache.cs
SqlitePersitentCache from Mapsui Extensions. This can cache urls too beside of tiles but I don't know if it has an offical mbtile layout.

Have you tried to add your cached mbtile to qgis and display it there. To test if it works?

**Reply by urossavelj** [reply]
@inforithmics I tried also with this now, adding to db works also with this, but I can't use this code than, because the database it creates is missing the metadata table and I get an error:
```
        var tileSource = new MbTilesTileSource(new SQLiteConnectionString(Path.Combine(FileSystem.AppDataDirectory, "myOfflineDb.mbtiles.sqlite")));

        _mapControl = new Mapsui.UI.Maui.MapControl();
        _mapControl.Map?.Layers.Add(new TileLayer(tileSource));
```

How do I set the source than, is there another way if you use the SqlitePersistentCache from Mapsui.Extensions?

**Reply by urossavelj** [reply]
I also tried loading the mbtiles files in qgis and it also doesn't work, so if I have to guess the downlaoding from Mapsui/BruTiels is not working correctly

**Comment by charlenni** [comment]
Did you turn logging on as suggested, normally is the first answer to such a problem? Second would be, are you sure, that you create a valid MBTiles file? Do you add all tables? Are all the data in the files are correct? Do you update the field "bounds"?

To say, that a well tested file work, but your own file doesn't work on two different system is the fault of Mapsui is a little bit strange ;-)

**Reply by urossavelj** [reply]
Well, I used the code from MapsUI website and Brutile Github. It's not like I invented it and documentation for this things is almost non existent, this is also why I'm asking here for help to figure out why is it not working as I can't find a working example online.

If you check on Brutile Github, this is exactly what I'm doing: https://github.com/BruTile/BruTile: 
```
Console.WriteLine("Show tile info");
foreach (var tileInfo in tileInfos)
{
    var tile = await tileSource.GetTileAsync(tileInfo);

    Console.WriteLine(
        $"tile col: {tileInfo.Index.Col}, " +
        $"tile row: {tileInfo.Index.Row}, " +
        $"tile level: {tileInfo.Index.Level} , " +
        $"tile size {tile.Length}");

      _cache.Add(tileInfo.Index, tile);
}
```
_cache is     `static SqlitePersistentCache _cache;`
and than adding each fetched tile to sqlite.

Than for displaying this is the only thing I do:

```
        var tileSource = new MbTilesTileSource(new SQLiteConnectionString(Path.Combine(FileSystem.AppDataDirectory, "myOfflineDb.mbtiles.sqlite")));

        _mapControl = new Mapsui.UI.Maui.MapControl();
        _mapControl.Map?.Layers.Add(new TileLayer(tileSource));
            
        Content = _mapControl;

        _mapControl.Map.Navigator.ZoomToLevel(8);
```

By using SqlitePersistentCache  cache and not MbTileCache, just the Tile table is created, no metadata. If I look at the "original" world.mbtiles file, the tables also have different names. Tile is called tiles and also additional metadata table(and also others). Can you tell me why the difference? If the "default" world.mbtiles file works I'm guessing that tiles, metadata is  correct and not "Tile" table?

**Comment by charlenni** [comment]
I assume, that you expect something, that isn't provided. A MBTiles cache or SqlitePersistanceCache create not a valid MBTiles file according to the specification (see [here](https://github.com/mapbox/mbtiles-spec/blob/master/1.1/spec.md)). So, if you want to create a valid MBTiles file, you have to do this by your own. Use the same table names according to the specification and fill all fields correct.

And, last, but not least, read the copyrights of the most map providers, that don't allow saving tiles for offline use ;-)

**Reply by urossavelj** [reply]
Yeah, I read the terms :)

Ah, it makes sense than, I was sure it makes a valid db file.
